pkg/tls/tls_test.go

```diff
@@ -0,0 +1,91 @@
+package tls
+
+import (
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"fmt"
+	"math/big"
+	"testing"
+	"time"
+)
+
+// x509 expects quite a large number so 0xDEADBEEF will not cut it
+func getSerialNumber() (*big.Int, error) {
+	serialNumber, err := rand.Int(rand.Reader, new(big.Int).Lsh(big.NewInt(1), 128))
+	if err != nil {
+		fmt.Errorf("failed to create cert serial number")
+	}
+	return serialNumber, nil
+}
+func getTemplate() (*x509.Certificate, error) {
+	now := time.Now()
+	SN, err := getSerialNumber()
+	if err != nil {
+		fmt.Errorf("Failed to create cert serial number")
+	}
+
+	template := &x509.Certificate{
+		SerialNumber:          SN,
+		NotBefore:             now,
+		NotAfter:              now.AddDate(5, 0, 0),
+		Subject:               pkix.Name{CommonName: cn},
+		BasicConstraintsValid: true,
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		IsCA:                  true,
+	}
+	return template, nil
+}
+
+var cn = "dummy-test-certificate.local"
+
+//func NewPrivateKey() (*rsa.PrivateKey, error) {
+func TestNewPrivateKey(t *testing.T) {
+	newPK, err := NewPrivateKey()
+	if err != nil {
+		t.Errorf("failed to create private key")
+	}
+	if newPK.Validate() != nil {
+		t.Errorf("Failed to validate private key")
+	}
+}
+
+//func newCert(key *rsa.PrivateKey, template *x509.Certificate,
+//signingkey *rsa.PrivateKey, signingcert *x509.Certificate)
+//(*x509.Certificate, error) {
+func testSelfSignedNewCert(t *testing.T) {
+	newPK, err := NewPrivateKey()
+	if err != nil {
+		t.Errorf("failed to create private key")
+	}
+	template, err := getTemplate()
+	if err != nil {
+		t.Errorf("failed to create template for Cert")
+	}
+	testCert, err := newCert(newPK, template, nil, nil)
+	if err != nil {
+		t.Errorf("failed to create test self signed certificate")
+	}
+	opts := x509.VerifyOptions{DNSName: cn}
+	if _, err := testCert.Verify(opts); err != nil {
+		t.Errorf("Failed to verify created self signed certificate")
+	}
+}
+
+//func NewCA(cn string) (*rsa.PrivateKey, *x509.Certificate, error) {
+func testNewCA(t *testing.T) {
+	pk, cert, err := NewCA(cn)
+	if err != nil {
+		t.Errorf("failed to create new CA")
+	}
+	if pk.Validate() != nil {
+		t.Errorf("failed to validate RSA Private key")
+	}
+	opts := x509.VerifyOptions{DNSName: cn}
+	if _, err := cert.Verify(opts); err != nil {
+		t.Errorf("Failed to verify created self signed certificate")
+	}
+}
```

Comment:
if you're going to do this, I think your approach here ends up being to zero out the parts of the returned cert which are dependent on time or the private key, then compare whatever is left to a known good value.  Not pretty.

Reply:
Then the same for `TestNewCert()`.